### PR TITLE
[SOFT-299] Add Gpio abstraction to adc library

### PR DESCRIPTION
**Pull request by mitchellostler**
Alright, back at it with another try. This time, the adcChannel defs are the all the same, and I built on top of the library with Gpio Defs. Will check results of PR check but should be all good. 

Overview of changes:
Adc.h
- Added wrappers for all current functions so that they can be called with gpioAddresses 
- Added compound literal gpioAddresses for adc internal channels (REF, TEMP, BAT)
- Added new Gpio AdcPinCallback typedef to allow funcs with gpios as params to be registered with register_callback_pin func
Adc.c
- Added new static AdcPinInterrupt array to keep track of registered callbacks and readings, ensured that although registered callbacks may exist in either AdcPinInterrupt or AdcInterrupt array, only one can be registered per channel
- Adjusted logic of ADC1_COMP_IRQHandler() to check which type of callback is registered, so that it can call with the right params
- Added x86 functionality
Test_adc.c
- Added tests for new functions using same logic as those used for adcchannel functions

### CODE REVIEW

File: libraries/ms-common/test/test_adc.c [hunk 1]
```diff
@@ -239,3 +275,121 @@ void test_adc_mock_reading() {
...
+  adc_init(ADC_MODE_SINGLE);
+  adc_set_channel_pin(ADC_CHANNEL_TEMP_PIN, true);
+
+  uint16_t reading;
+  adc_read_converted_pin(ADC_CHANNEL_TEMP_PIN, &reading);
+
+  // ensure value is within reason
+  TEST_ASSERT_TRUE(ADC_INVALID_UNDER_TEMP < reading);
+  TEST_ASSERT_TRUE(ADC_INVALID_OVER_TEMP > reading);
+}
```
Comment (copperium) [inline libraries/ms-common/test/test_adc.c:395]: This test won't be necessary if/when you remove `ADC_CHANNEL_TEMP_PIN` etc

**Review by copperium — CHANGES_REQUESTED** [review]
I'm not sure if we should be adding this much complexity to the module to maintain two options for the callback function, one taking a GpioAddress and one taking an AdcChannel. The callback feature isn't used at all outside test_adc.c in this codebase, and in the MSXII codebase it was only used in a demo "adc_driver" project. We could just say that to use the callback feature you'll have to use the legacy AdcChannel. Thoughts @g-jessmuir?

I also think it's worth supporting the legacy AdcChannel functions at least for `ADC_CHANNEL_TEMP/REF/BAT`, since I really don't like the compound literal GpioAddress hack to shoehorn ADC channel values into a GpioAddress - it's confusing and bug-prone.

Overall, great job Mitchell! I can tell you put a lot of work into improving adc.

File: libraries/ms-common/test/test_adc.c [hunk 2]
```diff
@@ -224,18 +244,109 @@ void test_adc_read_temp() {
   TEST_ASSERT_TRUE(ADC_INVALID_OVER_TEMP > reading);
 }
 
-// test to help with other tests
-void test_adc_mock_reading() {
```
Comment (copperium) [inline libraries/ms-common/test/test_adc.c:228]: Why did you remove this test?

**Review by copperium — CHANGES_REQUESTED** [review]
Just a couple nitpicks and things I missed last time, should be ready to merge after these updates

File: libraries/ms-common/test/test_adc.c [hunk 3]
```diff
@@ -224,18 +244,109 @@ void test_adc_read_temp() {
...
+  TEST_ASSERT_TRUE(s_pin_callback_ran);
+  TEST_ASSERT_TRUE(s_pin_callback_runs > 0);
+  TEST_ASSERT_TRUE(reading < 4095);
+
+  TEST_ASSERT_FALSE(s_callback_ran);
+}
+
+void test_pin_continuous() {
+  s_callback_runs = 0;
+  s_callback_ran = false;
```
[the remark this answers is not in the document]
Comment (mitchellostler) [reply libraries/ms-common/test/test_adc.c:312]: It is incremented in prv_rx_callback which is called by other tests as well. This means that I need to reset it before i run my test no?

File: libraries/ms-common/src/stm32f0xx/adc.c [hunk 4]
```diff
@@ -261,3 +280,31 @@ void ADC1_COMP_IRQHandler() {
...
+  AdcChannel channel;
+  adc_get_channel(address, &channel);
+  return adc_set_channel(channel, new_state);
+}
+
+StatusCode adc_register_callback_pin(GpioAddress address, AdcPinCallback callback, void *context) {
+  AdcChannel adc_channel;
+  adc_get_channel(address, &adc_channel);
+  status_ok_or_return(adc_register_callback(adc_channel, NULL, context));
+  s_adc_interrupts[adc_channel].pin_callback = callback;
```
[the remark this answers is not in the document]
Comment (mitchellostler) [reply libraries/ms-common/src/stm32f0xx/adc.c:296]: added prv_reset_channel() which sets all params of adcInterrupt struct to zero

[the remark this answers is not in the document]
Comment (mitchellostler) [reply libraries/ms-common/src/stm32f0xx/adc.c:296]: Also created prv_check_channel() function to replace same logic used in almost all the adcChannel funcs, and to allow gpio_pin_register_callback to run this check as well without adding redundant code

**Review by copperium — APPROVED** [review]

**Pull request merged**